File: Tareas_Go/ListaEnlazadaDoble.go
package main

import (
	"fmt"
)

/*Tipo de dato Persona*/
type Persona struct {
	nombre string
	edad int
	siguiente *Persona
	anterior *Persona
}

/*Tipo de dato ListaPersonas*/
type ListaPersonas struct {
	inicio_lista *Persona
	fin_lista *Persona
}

/*Nuevo objeto Persona*/
func nuevaPersona() *Persona {
	return new(Persona)
}

/*Nuevo objeto ListaPersonas*/
func nuevaListaPersonas() *ListaPersonas {
	return new(ListaPersonas)
}

/*Imprime los atributos de objeto Persona*/
func imprimirPersona(persona *Persona) {
	if persona != nil {
		fmt.Println("Nombre: ", persona.nombre, "Edad: ", persona.edad)	
	}
}

/*Devuelve verdadero si la lista esta vacia*/
func isListaVacia(listaPersonas *ListaPersonas) bool{
	
	if listaPersonas.inicio_lista == nil {
		fmt.Println("La lista esta vacia")
		return true
	}
	return false
}

/*Agrega un item elemento a la lista personas*/
func agregarPersonaLista(listaPersonas *ListaPersonas, persona *Persona) {

	persona.anterior = nil
	persona.siguiente = nil

	if listaPersonas.inicio_lista == nil {
		listaPersonas.inicio_lista = persona
	} else {
		persona.anterior = listaPersonas.fin_lista
		listaPersonas.fin_lista.siguiente = persona
	}

	listaPersonas.fin_lista = persona
	fmt.Println("Registro Agregado")
}

func buscarPersona(listaPersonas *ListaPersonas) *Persona{
	nodo_temporal := listaPersonas.fin_lista
	nombre := ingresarNombre()

	if isListaVacia(listaPersonas) {
		return nil
	}

	for  nodo_temporal !=  nil {
		if nodo_temporal.nombre == nombre {
			return nodo_temporal
		}
		nodo_temporal = nodo_temporal.anterior
	}
	
	fmt.Println("Valor no encontrado.")
	return nil
}

func eliminarPersona(listaPersonas *ListaPersonas) {

	nodo_temporal := listaPersonas.inicio_lista
	nombre := ingresarNombre()

	if isListaVacia(listaPersonas) {
		return
	}

	if listaPersonas.inicio_lista.nombre == nombre {
		listaPersonas.inicio_lista = listaPersonas.inicio_lista.siguiente
		return
	}

	for nodo_temporal.siguiente != nil {
		if nodo_temporal.siguiente.nombre == nombre {
			nodo_temporal.siguiente = nodo_temporal.siguiente.siguiente
			return
		}
		nodo_temporal = nodo_temporal.siguiente
	} 
}

/*Imprime todos los items de  ListaPersonas <ASC> <DES> de parametro*/
func imprimirLista(listaPersonas *ListaPersonas, modo string) {

	if isListaVacia(listaPersonas) {
		return
	}
	
	if modo == "DES" {
		nodo_temporal := listaPersonas.inicio_lista
		
		for nodo_temporal != nil {
			imprimirPersona(nodo_temporal)
			nodo_temporal = nodo_temporal.siguiente
		}
		return
	} 

	if modo == "ASC" {
		nodo_temporal := listaPersonas.fin_lista
		
		for nodo_temporal != nil {
			imprimirPersona(nodo_temporal)
			nodo_temporal = nodo_temporal.anterior
		}
	}
}

func imprimirArreglo(arregloPersonas []Persona) {
	
	for i := 0; i < len(arregloPersonas); i++ {
		imprimirPersona(&arregloPersonas[i])
	}
}

func tamanioLista(listaPersonas *ListaPersonas) int {
	_tamanio := 0
	nodo_temporal := listaPersonas.inicio_lista

	if isListaVacia(listaPersonas) {
		return _tamanio
	}

	for nodo_temporal != nil {
		_tamanio++
		nodo_temporal = nodo_temporal.siguiente
	}

	return _tamanio
}

/*Convierte la listaPersonas en un arreglo del mismo tipo*/
func convertirEnArreglo(listaPersonas *ListaPersonas) []Persona {
	nodo_temporal := listaPersonas.inicio_lista
	
	if isListaVacia(listaPersonas) {
		return nil
	}

	_tamanio := tamanioLista(listaPersonas)
	arregloPersonas := make([]Persona, _tamanio)

	for i := 0; i < _tamanio; i++ {
		arregloPersonas[i] = *nodo_temporal
		nodo_temporal = nodo_temporal.siguiente
	}

	fmt.Print("\n\tLa lista se ha convertido a un arreglo. \n")

	return arregloPersonas
}

/*Recibe los datos ingresados por el teclado*/
func ingresarDatosAgregar() *Persona {

	persona := nuevaPersona()

	persona.nombre = ingresarNombre()

	fmt.Print("Ingrese Edad: ")
	fmt.Scanln(&persona.edad)

	return persona
}

/*Recibe una cadena del teclado*/
func ingresarNombre() string {

	var nombre string

	fmt.Print("Ingrese Nombre: ")
	fmt.Scanln(&nombre)

	return nombre
}

func main() {
	listaPersonas := nuevaListaPersonas()
	var arregloPersonas []Persona

	seguir := true

	for seguir {
		fmt.Println("-- Lista enlazada doble Personas --")
		fmt.Println("\t1-. Agregar")
		fmt.Println("\t2-. Buscar")
		fmt.Println("\t3-. Eliminar")
		fmt.Println("\t4-. Imprimir Lista")
		fmt.Println("\t5-. Pasar a Arreglo")
		fmt.Println("\t6-. Imprimir Arreglo")
		fmt.Print("Ingrese una Opcion: ")

		var opcion int 
		fmt.Scanln(&opcion)

		switch opcion {
		case 1:
			persona := ingresarDatosAgregar()
			agregarPersonaLista(listaPersonas, persona)
		case 2:
			persona := buscarPersona(listaPersonas)
			imprimirPersona(persona)
		case 3:
			eliminarPersona(listaPersonas)
		case 4:
			imprimirLista(listaPersonas, "DES")
		case 5:
			arregloPersonas = convertirEnArreglo(listaPersonas)
			imprimirArreglo(arregloPersonas)
		case 6:
			imprimirArreglo(arregloPersonas)
		default:
			seguir = false				
		}
	}
}

File: TareaPython/Kata21/ListaEnlazadaDoble.py
class Persona(object):

	def __init__(self, nombre, edad):
		self.nombre = nombre
		self.edad = edad
		self.siguiente = None
		self.anterior = None

	def getNombre(self):
		return self.nombre

	def getEdad(self):
		return self.edad

	def imprimir(self):
		print "Nombre: ", self.getNombre(), "\tEdad: ", self.getEdad()

class ListaPersonas(object):
	def __init__(self):
		self.inicio_lista = None
		self.fin_lista = None
		self.arreglo = []

	def isListaVacia(self):
		if self.inicio_lista == None:
			print("La lista esta vacia")
			return True

	def insertar(self, persona):
		persona.anterior = None
		persona.siguiente = None

		if self.inicio_lista == None:
			self.inicio_lista = persona
		else:
			persona.anterior = self.fin_lista
			self.fin_lista.siguiente = persona
		self.fin_lista = persona
		print("Registro Agregado")

	def buscar(self, persona):
		if self.isListaVacia():
			return

		nodo_temporal = self.fin_lista
		while nodo_temporal != None:
			if nodo_temporal.getNombre() == persona.getNombre():
				return nodo_temporal
			nodo_temporal = nodo_temporal.anterior

	#No olvidar corregir la funcion eliminar para LD
	def eliminar(self, persona):
		if self.isListaVacia():
			return

		if self.inicio_lista.nombre == persona.nombre:
			if self.inicio_lista.siguiente == None:
				self.inicio_lista = self.fin_lista = None
				return
			self.inicio_lista.siguiente.anterior = None
			self.inicio_lista = self.inicio_lista.siguiente
			return

		if self.fin_lista.nombre == persona.nombre:
			self.fin_lista.anterior.siguiente = None
			self.fin_lista = self.fin_lista.anterior
			return

		nodo_temporal = self.buscar(persona)

		if nodo_temporal != None:
			nodo_temporal.anterior.siguiente = nodo_temporal.siguiente
			nodo_temporal.siguiente.anterior = nodo_temporal.anterior

	def imprimir(self, modo):
		if self.isListaVacia():
			return

		nodo_temporal = self.inicio_lista
		
		if modo == "DES":
			nodo_temporal = self.inicio_lista

			while (nodo_temporal != None):
				nodo_temporal.imprimir()
				nodo_temporal = nodo_temporal.siguiente		
			return

		if modo == "ASC":
			nodo_temporal = self.fin_lista

			while nodo_temporal != None:
				nodo_temporal.imprimir()
				nodo_temporal = nodo_temporal.anterior

	def convertirArreglo(self):
		self.arreglo = []
		if self.isListaVacia():
			return

		nodo_temporal = self.inicio_lista
		while  nodo_temporal != None:
			self.arreglo.append(nodo_temporal)
			nodo_temporal = nodo_temporal.siguiente

		print("\n\tLa lista se ha convertido a un arreglo. \n")

	def imprimirArreglo(self):
		x = 0
		while x < len(self.arreglo):
			self.arreglo[x].imprimir()
			x += 1

def main():
	listaPersonas = ListaPersonas()

	p1 = Persona('Luis', 13)
	p2 = Persona('Jose', 33)
	p3 = Persona('Mario', 16)
	p4 = Persona('Martha', 86)
	p5 = Persona('Jesis', 56)

	listaPersonas.insertar(p1)
	listaPersonas.insertar(p2)
	listaPersonas.insertar(p3)
	listaPersonas.insertar(p4)
	listaPersonas.insertar(p5)

	detener = False

	while not detener:
		print("-- Lista enlazada simple Personas --")
		print("\t1-. Agregar")
		print("\t2-. Buscar")
		print("\t3-. Eliminar")
		print("\t4-. Imprimir Lista")
		print("\t5-. Pasar a Arreglo")
		print("\t6-. Imprimir Arreglo")
		opcion = input("Ingrese una Opcion: ")

		if opcion == 1:
			nombre = ingresarNombre()
			edad = ingresarEdad()
			persona = pasarParametros(nombre, edad)
			listaPersonas.insertar(persona)
		elif opcion == 2:
			nombre = ingresarNombre()
			persona = pasarParametros(nombre, 0)
			persona = listaPersonas.buscar(persona)
			if persona != None:
				persona.imprimir()
		elif opcion == 3:
			nombre = ingresarNombre()
			persona = pasarParametros(nombre, 0)
			listaPersonas.eliminar(persona)
		elif opcion == 4:
			listaPersonas.imprimir("ASC")
		elif opcion == 5:
			listaPersonas.convertirArreglo()
		elif opcion == 6:
			listaPersonas.imprimirArreglo()
		else:
			detener = True


def pasarParametros(nombre, edad):
	persona = Persona(nombre, edad)
	return persona

def ingresarNombre():
	nombre = raw_input("Ingrese Nombre: ")
	return nombre

def ingresarEdad():
	edad = input("Ingrese Edad: ")
	return edad

if __name__ == '__main__':
	main()